File: src/futurnal/cli/learning.py
"""Learning Progress CLI commands for experiential learning visualization.

Step 08: Frontend Intelligence Integration - Phase 6

Provides commands for:
- Getting learning progress metrics
- Getting pattern learning stats
- Getting quality gates status
- Recording document learning events

Research Foundation:
- RLHI: Reinforcement Learning from Human Interactions
- AgentFlow: Learning from user feedback
- Option B: Ghost frozen, learning via token priors
"""

import json
import logging
from typing import Optional, List

from typer import Typer, Option, Argument

logger = logging.getLogger(__name__)

learning_app = Typer(help="Experiential learning progress commands")


def _get_persistent_pipeline():
    """Get the persistent ExperientialLearningPipeline singleton.

    Uses the singleton pattern to ensure state is preserved across CLI calls.
    Returns None if module not available.
    """
    try:
        from futurnal.learning.integration import get_persistent_pipeline
        return get_persistent_pipeline()
    except ImportError as e:
        logger.warning(f"Learning module not available: {e}")
        return None
    except Exception as e:
        logger.warning(f"Learning pipeline initialization failed: {e}")
        return None


@learning_app.command("progress")
def get_learning_progress(
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Get learning progress metrics and quality gates status.

    Examples:
        futurnal learning progress
        futurnal learning progress --json
    """
    try:
        # Get persistent learning pipeline (preserves state across calls)
        pipeline = _get_persistent_pipeline()

        if pipeline is not None:
            # Get stats from actual pipeline
            state = pipeline.state
            documents_processed = state.total_documents_processed
            success_rate = state.overall_success_rate

            # Quality progression
            if state.total_documents_processed > 0:
                quality_before = state.cumulative_quality_before / state.total_documents_processed
                quality_after = state.cumulative_quality_after / state.total_documents_processed
            else:
                quality_before = 0.0
                quality_after = 0.0
            improvement = quality_after - quality_before

            # Pattern learning stats from token store
            entity_priors = len(pipeline.token_store.entity_priors)
            relation_priors = len(pipeline.token_store.relation_priors)
            temporal_priors = len(pipeline.token_store.temporal_priors)

            # Quality gates
            ghost_frozen = True  # Always true per Option B
            improvement_threshold = 0.05
            meets_threshold = state.passes_quality_gate
        else:
            # Default/placeholder values when pipeline not available
            documents_processed = 0
            success_rate = 0.0
            quality_before = 0.0
            quality_after = 0.0
            improvement = 0.0
            entity_priors = 0
            relation_priors = 0
            temporal_priors = 0
            ghost_frozen = True
            improvement_threshold = 0.05
            meets_threshold = False

        if output_json:
            output = {
                "success": True,
                "documentsProcessed": documents_processed,
                "successRate": success_rate,
                "qualityProgression": {
                    "before": quality_before,
                    "after": quality_after,
                    "improvement": improvement,
                },
                "patternLearning": {
                    "entityPriors": entity_priors,
                    "relationPriors": relation_priors,
                    "temporalPriors": temporal_priors,
                },
                "qualityGates": {
                    "ghostFrozen": ghost_frozen,
                    "improvementThreshold": improvement_threshold,
                    "meetsThreshold": meets_threshold,
                },
            }
            print(json.dumps(output))
        else:
            print("Learning Progress:")
            print("-" * 40)
            print(f"\nDocuments Processed: {documents_processed}")
            print(f"Success Rate: {success_rate * 100:.1f}%")

            print("\nQuality Progression:")
            print(f"  Before: {quality_before:.2f}")
            print(f"  After: {quality_after:.2f}")
            print(f"  Improvement: {improvement:+.2f}")

            print("\nPattern Learning:")
            print(f"  Entity Priors: {entity_priors}")
            print(f"  Relation Priors: {relation_priors}")
            print(f"  Temporal Priors: {temporal_priors}")

            print("\nQuality Gates:")
            print(f"  Ghost Model Frozen: {ghost_frozen}")
            print(f"  Improvement Threshold: {improvement_threshold * 100:.0f}%")
            print(f"  Meets Threshold: {meets_threshold}")

    except Exception as e:
        logger.exception("Get learning progress failed")
        if output_json:
            print(json.dumps({"success": False, "error": str(e)}))
        else:
            print(f"Error: {e}")
        raise SystemExit(1)


@learning_app.command("record")
def record_document_learning(
    document_id: str = Argument(..., help="Unique identifier for the document"),
    content: str = Option("", "--content", "-c", help="Extracted text content"),
    source: str = Option("chat", "--source", "-s", help="Source of the document (chat, ingestion, etc.)"),
    content_type: str = Option("text", "--type", "-t", help="Content type (text, image, audio, document)"),
    success: bool = Option(True, "--success/--failure", help="Whether extraction was successful"),
    quality: Optional[float] = Option(None, "--quality", "-q", help="Quality score (0-1)"),
    entities: Optional[str] = Option(None, "--entities", "-e", help="Comma-separated entity types"),
    relations: Optional[str] = Option(None, "--relations", "-r", help="Comma-separated relation types"),
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Record a document processing event to the learning pipeline.

    Updates learning state and token priors, then persists to disk.
    Used by frontend to track chat attachments and processed documents.

    Examples:
        futurnal learning record doc_123 --content "Meeting notes..." --type text
        futurnal learning record img_456 --type image --entities "Image,Screenshot"
        futurnal learning record audio_789 --type audio --source chat --json
    """
    try:
        pipeline = _get_persistent_pipeline()

        if pipeline is None:
            error_msg = "Learning pipeline not available"
            if output_json:
                print(json.dumps({"success": False, "error": error_msg}))
            else:
                print(f"Error: {error_msg}")
            raise SystemExit(1)

        # Parse entity and relation types
        entity_types: Optional[List[str]] = None
        if entities:
            entity_types = [e.strip() for e in entities.split(",") if e.strip()]

        relation_types: Optional[List[str]] = None
        if relations:
            relation_types = [r.strip() for r in relations.split(",") if r.strip()]

        # Record document
        result = pipeline.record_document(
            document_id=document_id,
            content=content,
            source=source,
            content_type=content_type,
            success=success,
            quality_score=quality,
            entity_types=entity_types,
            relation_types=relation_types,
        )

        if output_json:
            output = {
                "success": True,
                "documentId": result["document_id"],
                "qualityScore": result["quality_score"],
                "totalDocuments": result["total_documents"],
                "overallSuccessRate": result["overall_success_rate"],
                "entityPriors": result["entity_priors"],
                "relationPriors": result["relation_priors"],
            }
            print(json.dumps(output))
        else:
            print(f"Recorded document: {document_id}")
            print(f"  Quality Score: {result['quality_score']:.2f}")
            print(f"  Total Documents: {result['total_documents']}")
            print(f"  Success Rate: {result['overall_success_rate'] * 100:.1f}%")
            print(f"  Entity Priors: {result['entity_priors']}")
            print(f"  Relation Priors: {result['relation_priors']}")

    except SystemExit:
        raise
    except Exception as e:
        logger.exception("Record document learning failed")
        if output_json:
            print(json.dumps({"success": False, "error": str(e)}))
        else:
            print(f"Error: {e}")
        raise SystemExit(1)


@learning_app.command("reset")
def reset_learning(
    confirm: bool = Option(False, "--yes", "-y", help="Confirm reset without prompt"),
    output_json: bool = Option(False, "--json", help="Output as JSON"),
) -> None:
    """Reset all learning state and token priors.

    WARNING: This clears all learned patterns and statistics.

    Examples:
        futurnal learning reset --yes
        futurnal learning reset --json --yes
    """
    try:
        if not confirm:
            print("This will reset all learning state and token priors.")
            print("Use --yes to confirm.")
            return

        pipeline = _get_persistent_pipeline()

        if pipeline is None:
            error_msg = "Learning pipeline not available"
            if output_json:
                print(json.dumps({"success": False, "error": error_msg}))
            else:
                print(f"Error: {error_msg}")
            raise SystemExit(1)

        # Reset learning
        result = pipeline.reset_learning()

        # Save the reset state
        pipeline.save_state()

        if output_json:
            output = {
                "success": True,
                "trajectoriesCleared": result["trajectories_cleared"],
                "priorsCleared": result["priors_cleared"],
                "stateReset": result["learning_state_reset"],
            }
            print(json.dumps(output))
        else:
            print("Learning state reset:")
            print(f"  Trajectories Cleared: {result['trajectories_cleared']}")
            print(f"  Priors Cleared: {result['priors_cleared']}")
            print(f"  State Reset: {result['learning_state_reset']}")

    except SystemExit:
        raise
    except Exception as e:
        logger.exception("Reset learning failed")
        if output_json:
            print(json.dumps({"success": False, "error": str(e)}))
        else:
            print(f"Error: {e}")
        raise SystemExit(1)

File: desktop/src-tauri/src/commands/learning.rs
//! Learning Progress IPC commands.
//!
//! Step 08: Frontend Intelligence Integration - Phase 6
//!
//! Research Foundation:
//! - RLHI: Reinforcement Learning from Human Interactions
//! - AgentFlow: Learning from user feedback
//! - Option B: Ghost frozen, learning via token priors
//!
//! Provides experiential learning progress metrics and document recording.

use serde::{Deserialize, Serialize};
use tauri::command;
use uuid::Uuid;

/// Quality progression metrics.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QualityProgression {
    pub before: f64,
    pub after: f64,
    pub improvement: f64,
}

/// Pattern learning statistics.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PatternLearning {
    pub entity_priors: i32,
    pub relation_priors: i32,
    pub temporal_priors: i32,
}

/// Quality gates status.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QualityGates {
    pub ghost_frozen: bool,
    pub improvement_threshold: f64,
    pub meets_threshold: bool,
}

/// Response from learning progress command.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LearningProgressResponse {
    pub success: bool,
    pub documents_processed: i32,
    pub success_rate: f64,
    pub quality_progression: QualityProgression,
    pub pattern_learning: PatternLearning,
    pub quality_gates: QualityGates,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Get learning progress metrics.
///
/// Step 08: Learning Progress Indicator
///
/// Calls: `futurnal learning progress --json`
#[command]
pub async fn get_learning_progress() -> Result<LearningProgressResponse, String> {
    let args = vec!["learning", "progress", "--json"];

    let response: LearningProgressResponse = match crate::python::execute_cli(&args).await {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("Learning progress CLI failed: {}", e);
            return Ok(LearningProgressResponse {
                success: false,
                documents_processed: 0,
                success_rate: 0.0,
                quality_progression: QualityProgression {
                    before: 0.0,
                    after: 0.0,
                    improvement: 0.0,
                },
                pattern_learning: PatternLearning {
                    entity_priors: 0,
                    relation_priors: 0,
                    temporal_priors: 0,
                },
                quality_gates: QualityGates {
                    ghost_frozen: true,
                    improvement_threshold: 0.05,
                    meets_threshold: false,
                },
                error: Some(format!("Learning progress failed: {}", e)),
            });
        }
    };

    log::info!(
        "Retrieved learning progress: {} documents, {:.1}% success rate",
        response.documents_processed,
        response.success_rate * 100.0
    );

    Ok(response)
}

/// Request payload for recording document learning.
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RecordDocumentRequest {
    /// Content extracted from the document.
    pub content: String,
    /// Source of the document (chat, ingestion, etc.).
    #[serde(default = "default_source")]
    pub source: String,
    /// Content type (text, image, audio, document).
    #[serde(default = "default_content_type")]
    pub content_type: String,
    /// Whether extraction was successful.
    #[serde(default = "default_success")]
    pub success: bool,
    /// Optional quality score (0-1).
    pub quality_score: Option<f64>,
    /// Optional entity types discovered.
    pub entity_types: Option<Vec<String>>,
    /// Optional relation types discovered.
    pub relation_types: Option<Vec<String>>,
}

fn default_source() -> String {
    "chat".to_string()
}

fn default_content_type() -> String {
    "text".to_string()
}

fn default_success() -> bool {
    true
}

/// Response from record document learning command.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RecordDocumentResponse {
    pub success: bool,
    pub document_id: String,
    pub quality_score: f64,
    pub total_documents: i32,
    pub overall_success_rate: f64,
    pub entity_priors: i32,
    pub relation_priors: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Record a document processing event to the learning pipeline.
///
/// Step 08: Full Learning Integration
///
/// Updates learning state and token priors, then persists to disk.
/// Called by frontend when processing chat attachments.
///
/// Calls: `futurnal learning record <document_id> --json [options]`
#[command]
pub async fn record_document_learning(
    request: RecordDocumentRequest,
) -> Result<RecordDocumentResponse, String> {
    // Generate unique document ID
    let document_id = format!("doc_{}", Uuid::new_v4().to_string().replace("-", "")[..12].to_string());

    let mut args = vec![
        "learning".to_string(),
        "record".to_string(),
        document_id.clone(),
        "--json".to_string(),
        "--source".to_string(),
        request.source.clone(),
        "--type".to_string(),
        request.content_type.clone(),
    ];

    // Add content if provided
    if !request.content.is_empty() {
        args.push("--content".to_string());
        args.push(request.content.clone());
    }

    // Add success/failure flag
    if !request.success {
        args.push("--failure".to_string());
    }

    // Add quality score if provided
    if let Some(quality) = request.quality_score {
        args.push("--quality".to_string());
        args.push(format!("{:.2}", quality));
    }

    // Add entity types if provided
    if let Some(entities) = &request.entity_types {
        if !entities.is_empty() {
            args.push("--entities".to_string());
            args.push(entities.join(","));
        }
    }

    // Add relation types if provided
    if let Some(relations) = &request.relation_types {
        if !relations.is_empty() {
            args.push("--relations".to_string());
            args.push(relations.join(","));
        }
    }

    // Convert to &str for CLI execution
    let args_ref: Vec<&str> = args.iter().map(|s| s.as_str()).collect();

    let response: RecordDocumentResponse = match crate::python::execute_cli(&args_ref).await {
        Ok(resp) => resp,
        Err(e) => {
            log::warn!("Record document learning CLI failed: {}", e);
            return Ok(RecordDocumentResponse {
                success: false,
                document_id,
                quality_score: 0.0,
                total_documents: 0,
                overall_success_rate: 0.0,
                entity_priors: 0,
                relation_priors: 0,
                error: Some(format!("Record document learning failed: {}", e)),
            });
        }
    };

    log::info!(
        "Recorded document {}: quality={:.2}, total={}",
        response.document_id,
        response.quality_score,
        response.total_documents
    );

    Ok(response)
}
